src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -2280,6 +2280,23 @@ private static void RegisterSerializationProviders(List<TypeInfo> providerTypes)
                         logger.Error(ErrorCode.SerMgr_ErrorLoadingAssemblyTypes, "Failed to create instance of type: " + type.FullName, exception);
                     }
                 });
+
+            // for the outside runtime client, a type may not be registered in the serialization manager before attempting to deserialize
+            // which ends up causing deserialization to fail.  to fix this, we need to register the type if it's supported by any serializer.
+            Action<Assembly> registerAction = 
+                assembly =>
+                {
+                    foreach (var type in assembly.GetTypes().Where(
+                        type => 
+                        (type.IsClass || type.IsValueType) && !type.IsAbstract &&
+                        externalSerializers.Any(ser => ser.IsSupportedType(type))))
+                    {
+                        SerializationManager.Register(type);
+                    }
+                };
+
+            AppDomain.CurrentDomain.AssemblyLoad += (o, e) => registerAction(e.LoadedAssembly);
+            Array.ForEach(AppDomain.CurrentDomain.GetAssemblies(), registerAction);
```

Comment:
This seems useful and might be related to some serialization issues we have been seeing in our tests, that prompt use to do workarounds such as calling `SerializationManager.InitializeForTesting()` before starting the cluster (although I'm not entirely sure if it's the same root cause).
Nevertheless this seems overly aggressive adding every single type that is loaded and can be serialized... can you or someone think of an approach that is a little bit more lazy than this?